sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java

```diff
@@ -0,0 +1,508 @@
+      this(longValue, null);
+    }
+
+    private FormattedItemValue(String longValue, String shortValue) {
+      this.longValue = longValue;
+      this.shortValue = shortValue;
+    }
+
+    private String getLongValue () {
+      return this.longValue;
+    }
+
+    private String getShortValue() {
+      return this.shortValue;
+    }
+  }
+
+  private static class InternalBuilder implements ItemBuilder {
+    private final Map<Identifier, Item> entries;
+    private final Set<Object> visited;
+
+    private Class<?> latestNs;
+    private Item latestItem;
+    private Identifier latestIdentifier;
+
+    private InternalBuilder() {
+      this.entries = Maps.newHashMap();
+      this.visited = Sets.newIdentityHashSet();
+    }
+
+    private static InternalBuilder forRoot(HasDisplayData instance) {
+      InternalBuilder builder = new InternalBuilder();
+      builder.include(instance);
+      return builder;
+    }
+
+    @Override
+    public Builder include(HasDisplayData subComponent) {
+      checkNotNull(subComponent);
+      boolean newComponent = visited.add(subComponent);
+      if (newComponent) {
+        Class prevNs = this.latestNs;
+        this.latestNs = subComponent.getClass();
+        subComponent.populateDisplayData(this);
+        this.latestNs = prevNs;
+      }
+
+      return this;
+    }
+
+    @Override
+    public ItemBuilder add(String key, String value) {
+      checkNotNull(value);
+      return addItem(key, Type.STRING, value);
+    }
+
+    @Override
+    public ItemBuilder add(String key, long value) {
+      return addItem(key, Type.INTEGER, value);
+    }
+
+    @Override
+    public ItemBuilder add(String key, double value) {
+      return addItem(key, Type.FLOAT, value);
+    }
+
+    @Override
+    public ItemBuilder add(String key, Instant value) {
+      checkNotNull(value);
+      return addItem(key, Type.TIMESTAMP, value);
+    }
+
+    @Override
+    public ItemBuilder add(String key, Duration value) {
+      checkNotNull(value);
+      return addItem(key, Type.DURATION, value);
+    }
+
+    @Override
+    public ItemBuilder add(String key, Class<?> value) {
+      checkNotNull(value);
+      return addItem(key, Type.JAVA_CLASS, value);
+    }
+
+    private <T> ItemBuilder addItem(String key, Type type, T value) {
+      checkNotNull(key);
+      checkArgument(!key.isEmpty());
+
+      Identifier id = Identifier.of(latestNs, key);
+      if (entries.containsKey(id)) {
+        throw new IllegalArgumentException("DisplayData key already exists. All display data "
+          + "for a component must be registered with a unique key.\nKey: " + id);
+      }
+      Item item = Item.create(id.getNamespace(), key, type, value);
+      entries.put(id, item);
+
+      latestItem = item;
+      latestIdentifier = id;
+
+      return this;
+    }
+
+    @Override
+    public ItemBuilder withLabel(String label) {
+      Item newItem = latestItem.withLabel(label);
```

Comment:
Interesting idea. Both ways would work; I'd prefer to settle on one syntax for simplicity.

I like the current syntax because it's fluent and should be discoverable and easy-to-use. Also, an implementer only has to interact with the DisplayData.Builder-- it never needs to look at the constructed DisplayData or Items.